test/framework/src/main/java/org/elasticsearch/index/mapper/BlockLoaderTestCase.java

```diff
@@ -0,0 +1,166 @@
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.tests.index.RandomIndexWriter;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.fieldvisitor.StoredFieldLoader;
+import org.elasticsearch.logsdb.datageneration.DataGeneratorSpecification;
+import org.elasticsearch.logsdb.datageneration.FieldDataGenerator;
+import org.elasticsearch.logsdb.datageneration.FieldType;
+import org.elasticsearch.logsdb.datageneration.MappingGenerator;
+import org.elasticsearch.logsdb.datageneration.Template;
+import org.elasticsearch.logsdb.datageneration.datasource.DataSourceHandler;
+import org.elasticsearch.logsdb.datageneration.datasource.DataSourceRequest;
+import org.elasticsearch.logsdb.datageneration.datasource.DataSourceResponse;
+import org.elasticsearch.search.fetch.StoredFieldsSpec;
+import org.elasticsearch.search.lookup.SearchLookup;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class BlockLoaderTestCase extends MapperServiceTestCase {
+    private final String fieldName;
+    private final Template template;
+    private final MappingGenerator mappingGenerator;
+    private final FieldDataGenerator generator;
+
+    protected BlockLoaderTestCase(FieldType fieldType) {
+        this.fieldName = randomAlphaOfLengthBetween(5, 10);
+
+        // Disable all dynamic mapping
+        var specification = DataGeneratorSpecification.builder()
+            .withFullyDynamicMapping(false)
+            .withDataSourceHandlers(List.of(new DataSourceHandler() {
+                @Override
+                public DataSourceResponse.DynamicMappingGenerator handle(DataSourceRequest.DynamicMappingGenerator request) {
+                    return new DataSourceResponse.DynamicMappingGenerator(isObject -> false);
+                }
+            }))
+            .build();
+
+        this.template = new Template(Map.of(fieldName, new Template.Leaf(fieldName, fieldType)));
+        this.mappingGenerator = new MappingGenerator(specification);
+        this.generator = fieldType.generator(fieldName, specification.dataSource());
+    }
+
+    public void testBlockLoader() throws IOException {
+        var mapping = mappingGenerator.generate(template);
+        var mappingXContent = XContentBuilder.builder(XContentType.JSON.xContent()).map(mapping.raw());
+
+        var syntheticSource = randomBoolean();
+        var mapperService = syntheticSource ? createSytheticSourceMapperService(mappingXContent) : createMapperService(mappingXContent);
+
+        var fieldValue = generator.generateValue();
+
+        Object blockLoaderResult = setupAndInvokeBlockLoader(mapperService, fieldValue);
```

Comment:
Can we randomly inject extra fields and/or documents to mimic the issue in https://github.com/elastic/elasticsearch/pull/117792, but that's for follow-ups.

Reply:
We could but i wonder if we should cover this in more integration level tests.